{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "146829e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.mplot3d import Axes3D\n",
    "from mpl_toolkits.axes_grid1 import make_axes_locatable\n",
    "from scipy.spatial import ConvexHull, convex_hull_plot_2d \n",
    "\n",
    "from scipy.stats import entropy\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42ebb762",
   "metadata": {},
   "source": [
    "## Prepare data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1209abda",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq</th>\n",
       "      <th>source</th>\n",
       "      <th>ShortName</th>\n",
       "      <th>DomainName</th>\n",
       "      <th>TranscriptID</th>\n",
       "      <th>Protein</th>\n",
       "      <th>DomainNo</th>\n",
       "      <th>ID</th>\n",
       "      <th>Species</th>\n",
       "      <th>Phylogeny</th>\n",
       "      <th>Sequences</th>\n",
       "      <th>Sequences_unaligned</th>\n",
       "      <th>MMD_z0</th>\n",
       "      <th>MMD_z1</th>\n",
       "      <th>MMD_z2</th>\n",
       "      <th>header_twist</th>\n",
       "      <th>norm_RE</th>\n",
       "      <th>orthologous_group</th>\n",
       "      <th>PhyGroup</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>jgi</td>\n",
       "      <td>Torde1</td>\n",
       "      <td>TDEL_0F02380_domain_number[1],jgi</td>\n",
       "      <td>3708.0</td>\n",
       "      <td>Equilibrative nucleoside transporter protein</td>\n",
       "      <td>1</td>\n",
       "      <td>4950.0</td>\n",
       "      <td>Torulaspora delbrueckii</td>\n",
       "      <td>cellular organisms; Eukaryota; Opisthokonta; F...</td>\n",
       "      <td>VFLGVYKALYDYEPQTTEELAIREDELLYLLEKSDEWWTVKKDEPV...</td>\n",
       "      <td>TVFLGVYKALYDYEPQTTEELAIREDELLYLLEKSDVDEWWTVKKR...</td>\n",
       "      <td>-0.212743</td>\n",
       "      <td>0.560967</td>\n",
       "      <td>-0.190477</td>\n",
       "      <td>4837_jgi||3708||Equilibrative nucleoside trans...</td>\n",
       "      <td>0.148125</td>\n",
       "      <td>NOG257967_1</td>\n",
       "      <td>Ascomycota</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>jgi</td>\n",
       "      <td>Torde1</td>\n",
       "      <td>TDEL_0F02380_domain_number[2],jgi</td>\n",
       "      <td>3708.0</td>\n",
       "      <td>Equilibrative nucleoside transporter protein</td>\n",
       "      <td>2</td>\n",
       "      <td>4950.0</td>\n",
       "      <td>Torulaspora delbrueckii</td>\n",
       "      <td>cellular organisms; Eukaryota; Opisthokonta; F...</td>\n",
       "      <td>PVINSVRALYDYDQNPDEELTFHENEEFDVYDDQDDWLLVQKTGAC...</td>\n",
       "      <td>APVINSVRALYDYDQAQNPDEELTFHENEEFDVYDDQDPDWLLVQQ...</td>\n",
       "      <td>-0.133021</td>\n",
       "      <td>1.238067</td>\n",
       "      <td>-0.209800</td>\n",
       "      <td>2997_jgi||3708||Equilibrative nucleoside trans...</td>\n",
       "      <td>0.076445</td>\n",
       "      <td>NOG257967_2</td>\n",
       "      <td>Ascomycota</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>jgi</td>\n",
       "      <td>Torde1</td>\n",
       "      <td>TDEL_0F02380_domain_number[3],jgi</td>\n",
       "      <td>3708.0</td>\n",
       "      <td>Equilibrative nucleoside transporter protein</td>\n",
       "      <td>3</td>\n",
       "      <td>4950.0</td>\n",
       "      <td>Torulaspora delbrueckii</td>\n",
       "      <td>cellular organisms; Eukaryota; Opisthokonta; F...</td>\n",
       "      <td>KARKEGKVLYDFIAESQDELSVKQGQTVYILNDKKDWWMCELSGQK...</td>\n",
       "      <td>SKARKEGKVLYDFIAESQDELSVKQGQTVYILNDKKSRDWWMCELV...</td>\n",
       "      <td>-1.025860</td>\n",
       "      <td>0.236221</td>\n",
       "      <td>-0.425561</td>\n",
       "      <td>4599_jgi||3708||Equilibrative nucleoside trans...</td>\n",
       "      <td>-0.175593</td>\n",
       "      <td>NOG257967_3</td>\n",
       "      <td>Ascomycota</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>jgi</td>\n",
       "      <td>Torde1</td>\n",
       "      <td>TDEL_0B05220_domain_number[1],jgi</td>\n",
       "      <td>1310.0</td>\n",
       "      <td>Cdc42-interacting protein CIP4</td>\n",
       "      <td>1</td>\n",
       "      <td>4950.0</td>\n",
       "      <td>Torulaspora delbrueckii</td>\n",
       "      <td>cellular organisms; Eukaryota; Opisthokonta; F...</td>\n",
       "      <td>TGSNKNKVLFAYSKQDADEVSISLRDLVALEAADGGWTKIKNTGET...</td>\n",
       "      <td>GTGSNKNKVLFAYSKQDADEVSISLRDLVALEAADTGSGWTKIKNL...</td>\n",
       "      <td>-0.778632</td>\n",
       "      <td>-0.124500</td>\n",
       "      <td>-0.325014</td>\n",
       "      <td>5418_jgi||1310||Cdc42-interacting protein CIP4...</td>\n",
       "      <td>0.119790</td>\n",
       "      <td>KOG3565_1</td>\n",
       "      <td>Ascomycota</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>jgi</td>\n",
       "      <td>Torde1</td>\n",
       "      <td>TDEL_0B05220_domain_number[2],jgi</td>\n",
       "      <td>1310.0</td>\n",
       "      <td>Cdc42-interacting protein CIP4</td>\n",
       "      <td>2</td>\n",
       "      <td>4950.0</td>\n",
       "      <td>Torulaspora delbrueckii</td>\n",
       "      <td>cellular organisms; Eukaryota; Opisthokonta; F...</td>\n",
       "      <td>MPTRTMQAQYDYEAQGDDELSLTPNDVVNVIRGDDGWTYGE-NGEK...</td>\n",
       "      <td>NMPTRTMQAQYDYEAQGDDELSLTPNDVVNVIRGDDGSGWTYGELN...</td>\n",
       "      <td>-0.155759</td>\n",
       "      <td>-0.226678</td>\n",
       "      <td>-0.148836</td>\n",
       "      <td>4838_jgi||1310||Cdc42-interacting protein CIP4...</td>\n",
       "      <td>-0.255056</td>\n",
       "      <td>KOG3565_2</td>\n",
       "      <td>Ascomycota</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   seq source ShortName                         DomainName  TranscriptID  \\\n",
       "0    0    jgi    Torde1  TDEL_0F02380_domain_number[1],jgi        3708.0   \n",
       "1    1    jgi    Torde1  TDEL_0F02380_domain_number[2],jgi        3708.0   \n",
       "2    2    jgi    Torde1  TDEL_0F02380_domain_number[3],jgi        3708.0   \n",
       "3    3    jgi    Torde1  TDEL_0B05220_domain_number[1],jgi        1310.0   \n",
       "4    4    jgi    Torde1  TDEL_0B05220_domain_number[2],jgi        1310.0   \n",
       "\n",
       "                                        Protein  DomainNo      ID  \\\n",
       "0  Equilibrative nucleoside transporter protein         1  4950.0   \n",
       "1  Equilibrative nucleoside transporter protein         2  4950.0   \n",
       "2  Equilibrative nucleoside transporter protein         3  4950.0   \n",
       "3                Cdc42-interacting protein CIP4         1  4950.0   \n",
       "4                Cdc42-interacting protein CIP4         2  4950.0   \n",
       "\n",
       "                   Species                                          Phylogeny  \\\n",
       "0  Torulaspora delbrueckii  cellular organisms; Eukaryota; Opisthokonta; F...   \n",
       "1  Torulaspora delbrueckii  cellular organisms; Eukaryota; Opisthokonta; F...   \n",
       "2  Torulaspora delbrueckii  cellular organisms; Eukaryota; Opisthokonta; F...   \n",
       "3  Torulaspora delbrueckii  cellular organisms; Eukaryota; Opisthokonta; F...   \n",
       "4  Torulaspora delbrueckii  cellular organisms; Eukaryota; Opisthokonta; F...   \n",
       "\n",
       "                                           Sequences  \\\n",
       "0  VFLGVYKALYDYEPQTTEELAIREDELLYLLEKSDEWWTVKKDEPV...   \n",
       "1  PVINSVRALYDYDQNPDEELTFHENEEFDVYDDQDDWLLVQKTGAC...   \n",
       "2  KARKEGKVLYDFIAESQDELSVKQGQTVYILNDKKDWWMCELSGQK...   \n",
       "3  TGSNKNKVLFAYSKQDADEVSISLRDLVALEAADGGWTKIKNTGET...   \n",
       "4  MPTRTMQAQYDYEAQGDDELSLTPNDVVNVIRGDDGWTYGE-NGEK...   \n",
       "\n",
       "                                 Sequences_unaligned    MMD_z0    MMD_z1  \\\n",
       "0  TVFLGVYKALYDYEPQTTEELAIREDELLYLLEKSDVDEWWTVKKR... -0.212743  0.560967   \n",
       "1  APVINSVRALYDYDQAQNPDEELTFHENEEFDVYDDQDPDWLLVQQ... -0.133021  1.238067   \n",
       "2  SKARKEGKVLYDFIAESQDELSVKQGQTVYILNDKKSRDWWMCELV... -1.025860  0.236221   \n",
       "3  GTGSNKNKVLFAYSKQDADEVSISLRDLVALEAADTGSGWTKIKNL... -0.778632 -0.124500   \n",
       "4  NMPTRTMQAQYDYEAQGDDELSLTPNDVVNVIRGDDGSGWTYGELN... -0.155759 -0.226678   \n",
       "\n",
       "     MMD_z2                                       header_twist   norm_RE  \\\n",
       "0 -0.190477  4837_jgi||3708||Equilibrative nucleoside trans...  0.148125   \n",
       "1 -0.209800  2997_jgi||3708||Equilibrative nucleoside trans...  0.076445   \n",
       "2 -0.425561  4599_jgi||3708||Equilibrative nucleoside trans... -0.175593   \n",
       "3 -0.325014  5418_jgi||1310||Cdc42-interacting protein CIP4...  0.119790   \n",
       "4 -0.148836  4838_jgi||1310||Cdc42-interacting protein CIP4... -0.255056   \n",
       "\n",
       "  orthologous_group    PhyGroup  \n",
       "0       NOG257967_1  Ascomycota  \n",
       "1       NOG257967_2  Ascomycota  \n",
       "2       NOG257967_3  Ascomycota  \n",
       "3         KOG3565_1  Ascomycota  \n",
       "4         KOG3565_2  Ascomycota  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Natural sequences\n",
    "merge_ds = pd.read_excel('./Taxonomy_and_convex_hull/data/20210908_natural_with_taxonomy.xls').iloc[:,1:]\n",
    "merge_ds.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "abbe19d5",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'nat_df' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/mn/43wl4v5s4tlddblhclx4dpyh0000gn/T/ipykernel_42784/2791486053.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# merge the two natural sh3 spreadsheets together\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;31m# note: there should be both the 'In training set' and 'MMD z..' columns\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m full_nat_df = nat_df.merge( \n\u001b[0m\u001b[1;32m      4\u001b[0m     \u001b[0mmerge_ds\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m     left_on='Sequence_aligned', right_on='Sequences').drop_duplicates(subset='Sequences_unaligned_x').reset_index(drop = True)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'nat_df' is not defined"
     ]
    }
   ],
   "source": [
    "# merge the two natural sh3 spreadsheets together \n",
    "# note: there should be both the 'In training set' and 'MMD z..' columns \n",
    "full_nat_df = nat_df.merge( \n",
    "    merge_ds,\n",
    "    left_on='Sequence_aligned', right_on='Sequences').drop_duplicates(subset='Sequences_unaligned_x').reset_index(drop = True)\n",
    "\n",
    "train_full_nat_df = full_nat_df[full_nat_df['In Training Set']==1]\n",
    "train_full_nat_df = train_full_nat_df.drop_duplicates(subset='Sequence_aligned').reset_index(drop = True)\n",
    "train_full_nat_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e2c0398",
   "metadata": {},
   "outputs": [],
   "source": [
    "# === functional sequences defined by r.e. > 0.5 ===\n",
    "func_train_full_nat_df = train_full_nat_df[train_full_nat_df.norm_RE > 0.5]\n",
    "# embeddings\n",
    "func_train_full_nat_z = func_train_full_nat_df[['MMD_z0', 'MMD_z1', 'MMD_z2']].values\n",
    "\n",
    "# === nonfunctional sequences definde by r.e. <= 0.5 ===\n",
    "nonfunc_train_full_nat_df = train_full_nat_df[~(train_full_nat_df.norm_RE > 0.5)]\n",
    "# embeddings\n",
    "nonfunc_train_full_nat_z = nonfunc_train_full_nat_df[['MMD_z0', 'MMD_z1', 'MMD_z2']].values\n",
    "\n",
    "func_train_full_nat_z.shape, nonfunc_train_full_nat_z.shape\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f32af37a",
   "metadata": {},
   "source": [
    "## Plot Figure 3D:\n",
    "\n",
    "Here, we are plotting all of the natural training sequences (5299). The pink points are 162 functional sequences (r.e.>0.5) and  5137 remaining embeddings are blue. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c728d6fb",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "plt.rcParams.update({'font.size': 12})\n",
    "plt.rcParams['xtick.labelsize'] = 12\n",
    "plt.rcParams.update({'font.family': 'Arial'})\n",
    "\n",
    "dsize = 1\n",
    "size =15\n",
    "lw=.5\n",
    "bins = np.linspace(-2.5,2.5,25)\n",
    "\n",
    "fig, axs = plt.subplots(1, 2, figsize = [6.0,3], sharex=\"col\", sharey=\"row\", dpi=300)\n",
    "\n",
    "\n",
    "#axs[0, 1].hist(mmd_nat_coord[good_nat,2], bins, orientation='horizontal', color = '#f47b7b')\n",
    "\n",
    "# functional latent embeddings\n",
    "axs[0].scatter(nonfunc_train_full_nat_z[:,0], nonfunc_train_full_nat_z[:,2]\n",
    "               ,s=size, c = 'cornflowerblue',alpha=.5)\n",
    "\n",
    "axs[0].scatter(func_train_full_nat_z[:,0], func_train_full_nat_z[:,2],\n",
    "              c = '#f47c7c', linewidth=lw,s = size,alpha=.5 )\n",
    "\n",
    "# nonfunctional latent embeddings\n",
    "axs[1].scatter(nonfunc_train_full_nat_z[:,1], nonfunc_train_full_nat_z[:,2]\n",
    "                                          ,s=size, c = 'cornflowerblue',alpha=.5)\n",
    "axs[1].scatter(func_train_full_nat_z[:,1], func_train_full_nat_z[:,2],\n",
    "              c = '#f47c7c', linewidth=lw,s = size,alpha=.5)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# subplot 0 ticks\n",
    "axs[0].set_xticks([-2.5, 0, 2.5])\n",
    "axs[0].set_yticks([-2, 0, 2])\n",
    "\n",
    "# subplot 1 ticks\n",
    "axs[1].set_xticks([-2.5, 0, 2.5])\n",
    "axs[1].set_yticks([-2, 0, 2])\n",
    "    \n",
    "axs[0].set_xlabel('Dim 1')\n",
    "axs[0].set_ylabel('Dim 3')\n",
    "axs[1].set_xlabel('Dim 2')\n",
    "\n",
    "\n",
    "plt.tight_layout(pad=.8)\n",
    "plt.savefig('./figures/Fig3D_natural_func_embeddings.svg', dpi = 300)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "27aacc73",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
